File: src/main/java/com/fouan/board/ZoneUtils.java
package com.fouan.board;

import com.fouan.utils.ListUtils;

import java.util.Comparator;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class ZoneUtils {

    public static final Comparator<Zone> compareZonePerNoise = Comparator.comparing(Zone::getNoise);

    public static List<Zone> getNoisiestZones(List<Zone> zones, boolean withSurvivors) {
        Set<Zone> matchingZones = zones.stream()
                .filter(zone -> !withSurvivors || zone.containsSurvivor())
                .collect(Collectors.toSet());

        return ListUtils.keepMax(matchingZones, compareZonePerNoise);
    }
}

File: src/main/java/com/fouan/old/board/ZoneUtils.kt
package com.fouan.old.board

object ZoneUtils {
    //    public static final Comparator<Zone> compareZonePerNoise = Comparator.comparing(Zone::getNoise);
    @JvmStatic
    fun getNoisiestZones(zones: List<Zone>, withSurvivors: Boolean): List<Zone> {
        val matchingZones = zones.filter { !withSurvivors || it.containsSurvivor() }
        val maxNoise = matchingZones.maxOfOrNull { it.noise }

        return maxNoise?.let {
            matchingZones.filter { zone: Zone -> zone.noise == it }
        } ?: emptyList()
    }
}
